### ISSUE TITLE
`deno compile` output can't dynamically import files that were not present at compilation time

### ISSUE THREAD

**Issue by alcuadrado** [denoland/deno#18327]
_(Creating this issue as a continuation of https://github.com/denoland/deno/issues/8655)_

As @dsherret commented in #8655, Deno 1.31 added support for [statically analyzable dynamic imports](https://deno.com/blog/v1.31#deno-compile-works-with-statically-analyzable-dynamic-imports), and Deno 1.32 will add support for web workers (https://github.com/denoland/deno/commit/090169cfbc6699486765b729d532b5b837210b12) and the ability to specify non-analyzable modules to include in the binary (https://github.com/denoland/deno/commit/b64ec7926831896f4e43b685891111409de45e85).

These are good improvements to `deno compile`, but they don't cover many use-cases described in #8655.

Here's an incomplete list of things that `deno compile`'s output can't do:

* [Config files written in javascript/typescript](https://github.com/denoland/deno/issues/8655#issuecomment-1437444576) 
* [User-defined plugins](https://github.com/denoland/deno/issues/8655#issuecomment-1247057354)
* [Inspect user supplied values (`graphgen inspect path/to/your/custom/setup.ts`)](https://github.com/denoland/deno/issues/8655#issuecomment-1247057354)
* [Writing interpreters of other languages](https://github.com/denoland/deno/issues/8655#issuecomment-1173616850)
* [Running user javascript files created by a deno-based compiler](https://github.com/denoland/deno/issues/8655)
* [Filesystem-based routing](https://github.com/denoland/deno/issues/8655#issuecomment-833422361)

**Comment by dojyorin** [comment]
I hope this issue will be resolved.
I want to load JSON modules config file at runtime with dynamic import.

**Comment by cowboyd** [comment]
In the particular use-case of importing at runtime to define a plugin, it would be nice if you could restrict the permissions that the module and its dependencies will have, perhaps even starting from zero. That way, you could have a process with expansive permissions safely load code that is sandboxed to a more restrictive security context.

For example, to load a plugin such that it can only read from its own plugin directory and nothing else:

```ts
async function loadPlugin(pluginName) {
  let home = Deno.env.get("HOME");
  let pluginDir = `${home}/.plugins/${pluginName}`
  return await import(module, { assert: { allowRead: pluginDir } });
}
```

**Comment by alcuadrado** [comment]
I believe that's not possible in Deno nor v8, @cowboyd. It's similar to [this issue](https://github.com/denoland/deno/issues/378). I agree that it would be awesome.

**Comment by johnspurlock** [comment]
> it would be nice if you could restrict the permissions that the module and its dependencies will have, perhaps even starting from zero.

@cowboyd I do this in [Denoflare](https://denoflare.dev/) using [permissionless Workers](https://docs.deno.com/runtime/manual/runtime/workers#specifying-worker-permissions), which is also useful for in-process hot-reloading - each Worker is a separate isolate inside the single Deno process that can be torn down.

In my case, I don't have to emulate anything in the Deno namespace since the user-provided scripts are intended to support running on a limited web-standard-only environment like Cloudflare Workers.  It would be trickier, but not impossible, if you had to proxy certain Deno ns calls too.

I've been waiting on `deno compile` support for dynamic imports (of blob: urls in my case) for a very long time now, would be so nice...

**Comment by dogukanakkaya** [comment]
I need this as well. My use-case is a GraphQL based API setup:
```ts
// "graphql-tools/load-files": "npm:@graphql-tools/load-files@7.0.0",
import { loadFiles } from "graphql-tools/load-files";

const typeDefFiles = await loadFiles(path.join(Deno.cwd(), "src/**/types.graphql"));
const resolverFiles = await loadFiles(path.join(Deno.cwd(), "src/**/resolvers.ts"));
```

`typeDefFiles` line is completely okay and working fine. But when it comes to `resolverFiles` I got this error:
> error: Uncaught (in promise) TypeError: Module not found: file:///.../src/graphql/resolvers.ts
            const fileExports = await requireMethod(path);

I assume Deno Deploy also compiles projects into executables before deployment because I got the same error there as well. Today I noticed we have an issue for this. [I already created one for Deno Deploy's repository](https://github.com/denoland/deploy_feedback/issues/518) but I think that can be closed in favor of this, if Deno Deploy also uses `compile`.

**Comment by jeff-hykin** [comment]
It would be nice if `deno compile` would give an error when it can't statically import a dynamic import instead of just waiting till runtime to throw. I was going insane trying to figure out how the following was possible, as I assumed that dynamic imports would either be fully supported or not supported at all. 

```
// works
import(`file:///Users/jeffhykin/.cache/nvs//rikudoeSage.allPackages.cache.js`)

// runtime error
const path2 = `file:///Users/jeffhykin/.cache/nvs//rikudoeSage.allPackages.cache.js`
import(path2)
```

**Comment by jeff-hykin** [comment]
@cowboyd  @dogukanakkaya @johnspurlock 

This will probably work well enough for you guys (a little-known feature of ES imports)

```js
const pathToLibray = `${Math.random()}`
const import = await import(`data:text/javascript;base64,${btoa(await Deno.readTextFile(pathToLibray))}`)
```

Works for me on Deno `1.31.1`, `1.33.1`, and (latest) `1.38.2`

**Comment by dsherret** [comment]
Can't really eror at compile time because the dynamic import error not resolving might be caught. If anything, it should warn, but that would be a bit of effort to implement.

At the moment it includes dynamic imports that are statically analyzable if they succeed to load. By that, it currently means any dynamic import that has the string literal or template literal (with no expressions/interpolation) as the argument to an `import(...)` call.

In the next minor release it will also identify situations such as the following with a template literal (so it will automatically embed all .js files found in sub_folder):

```ts
const mod = await import(`./sub_folder/${name}.js`);
```

See https://github.com/denoland/deno_graph/issues/275

**Comment by jeff-hykin** [comment]
> Can't really eror at compile time because the dynamic import error not resolving might be caught.

Ah, that makes sense. I can see offline tools doing that a lot and having a valid fallback.

> but that would be a bit of effort to implement.

Really? I mean, statically-import-a-dynamic-import sounds like a ton of work, definitely. But since thats already done, it seems really trivial to do `try { loadDyanmicImport() } catch() { console.warn() }`. Is there something I'm missing?

**Comment by dsherret** [comment]
> > but that would be a bit of effort to implement.
> 
> Really? `try { loadDyanmicImportAtCompileTime() } catch() { console.warn() }` I mean, statically-import-a-dynamic-import sounds like a ton of work (and so does that template literal situation). But adding the catch seems straightforward to me.
> 
> Is there something I'm missing?

I'm not sure what you mean. What I meant is that getting Deno to warn at compile time (not runtime) for dynamic imports that it couldn't analyze for deno_compile would be a bit of effort, though now that I think about it a bit more... it wouldn't be too bad after one of my [recent refactors](https://github.com/denoland/deno_graph/pull/332) that actually stores this information in the module analysis cache now.

**Comment by jeff-hykin** [comment]
> I'm not sure what you mean.

Somewhere in the code, while `deno compile` is running, it figures out that a dynamic import can be statically loaded right? ("dynamic imports that are statically analyzable if they succeed to load")

My thought is to simply change "if they succeed to load" to "if they succeed to load" `else console.warn()`

**Comment by 7flash** [comment]
In my compilable app, I have implemented function which is loading modules from separate github repository in runtime:

https://github.com/galaxydo/galaxy-desktop-app/blob/52e78c02e635ffa92b6c9ba50aa3be4f24f7ce53/main.ts#L77

But I understand it posses security risks for end user since I can modify those files independently, and looking for a better solution to embed those files at compile time.

**Comment by dsherret** [comment]
> > I'm not sure what you mean.
> 
> Somewhere in the code, while `deno compile` is running, it figures out that a dynamic import can be statically loaded right? ("dynamic imports that are statically analyzable if they succeed to load" in [comment](https://github.com/denoland/deno/issues/18327#issuecomment-1823582399))
> 
> My thought is to simply change "if they succeed to load" to "if (they succeed to load) {...} else { console.warn() }"

Ah, I see what you mean. I got confused because of the JS code. Short answer is no it's not that simple and it doesn't exactly work that way, but also it's not as difficult anymore due to the refactor I linked to in my previous comment.

**Comment by jeff-hykin** [comment]
## TLDR for anyone just arriving

This [community tool](https://github.com/ayoreis/import) seems like a solid workaround.


```js
import { dynamicImport } from 'https://deno.land/x/import/mod.ts';
await dynamicImport(Math.random() < 0.5 ? './foo.ts' : './bar.ts');
```

(Also thank you @7flash I only found that tool after looking through your code)

**Comment by davidbarratt** [comment]
I have a use case that I think is hugely beneficial, yet narrow enough to not allow _any_ dynamic import.

Drupal now has support for [single directory components](https://www.drupal.org/docs/develop/theming-drupal/using-single-directory-components) which are similar to a [React](https://react.dev/) component, but because it's in PHP it only supports [Twig](https://twig.symfony.com/). Support for React is [TBD](https://www.drupal.org/docs/develop/theming-drupal/using-single-directory-components/faq-frequently-asked-questions#s-can-i-use-my-react-components-within-sdc).

The way I've seen other projects like [Hypernova](https://github.com/airbnb/hypernova) handle this is by spinning up a Node.js server and sending `POST` requests to that server. This is obviously a lot for someone to setup and run.

A better way might be to have a binary that is effectively just a shim around [renderToReadableStream](https://react.dev/reference/react-dom/server/renderToReadableStream) (or whichever framework you prefer). Drupal already commonly uses executables like [ImageMagick](https://www.drupal.org/project/imagemagick). To me, having an executable that can render the JSX template into HTML is way better than having to ask the user to install Node.js and keep a server running on `localhost`.

The nice thing here though, is that the dynamically imported module has to be provided to the executable. For this use case, Deno could restrict dynamically imported modules to only those that are 1) found on the local filesystem and 2) specified as an argument to the binary.

I imagine the usage of a rendering binary would look something like this:
```sh
./render ./path/to/Component.tsx '{ "foo": "bar" }'
```
as in, the binary, the path to the component to render and then a JSON blob of props to pass to the component after it has been dynamically loaded.

Unless there is a better way to do this? I imagine many other non-JS systems (i.e. [WordPress](https://wordpress.org/), [MediaWiki](https://www.mediawiki.org/wiki/MediaWiki), [Shopify](https://www.shopify.com/), etc.) would benefit immensely by having binaries that can render JS components (from any framework). Unless this already exists?

**Comment by davidbarratt** [comment]
Not to get too much into creating solutions, but if you wanted to enforce having to pass in the dynamic import, you could use a new specifier like `deno:args:0` i.e. 
```ts
import Component from 'deno:args:0';
```
which would fail if it cannot be imported or
```ts
const { default: Component } = await import('deno:args:0'); 
```
which would allow the failure to be handled.

**Comment by Mutefish0** [comment]
> Can't really eror at compile time because the dynamic import error not resolving might be caught. If anything, it should warn, but that would be a bit of effort to implement.
> 
> At the moment it includes dynamic imports that are statically analyzable if they succeed to load. By that, it currently means any dynamic import that has the string literal or template literal (with no expressions/interpolation) as the argument to an `import(...)` call.
> 
> In the next minor release it will also identify situations such as the following with a template literal (so it will automatically embed all .js files found in sub_folder):
> 
> ```ts
> const mod = await import(`./sub_folder/${name}.js`);
> ```
> 
> See [denoland/deno_graph#275](https://github.com/denoland/deno_graph/issues/275)


It would be even better if we could identify the template arguments using specific environment variables like `Deno.build.target`.

For instance, with the following files:

- main.ts
```typescript
const data = await import(`./base64/${Deno.build.target}.ts`);
```
- base64
  - aarch64-apple-darwin.ts
  - x86_64-unknown-linux-gnu.ts

Running `deno compile --target aarch64-apple-darwin main.ts` will only include `base64/aarch64-apple-darwin.ts`, which can help reduce the final binary size :)

**Comment by 0xNF** [comment]
I've run into this issue.

>  deno --version
deno 2.0.4 (stable, release, x86_64-pc-windows-msvc)
v8 12.9.202.13-rusty
typescript 5.6.2

I'm using webworkers to load .ts modules that are on the local machine but not known at compile-time. It's basically a user addon system. It is of course running fine with `deno run`.  Hitting the code path where the import happens in the compiled version gives me this `ERR_MODULE_NOT_FOUND`.


using the dynamicImport link posted earlier in the tread just gives me vague `NotCapable` errors.

**Comment by maliknajjar** [comment]
I really need this feature for my CLI tool that gets a typescript configuration file

**Comment by chyzwar** [comment]
I extracted needed functionality from https://github.com/ayoreis/import
```ts
import { build, stop } from "https://deno.land/x/esbuild@v0.24.0/mod.js";
import { denoPlugins } from "@luca/esbuild-deno-loader";
import type { BuildOptions, Plugin } from "https://deno.land/x/esbuild@v0.24.0/mod.js";

/**
 * Below is a modified version of the import function from
 * @see https://github.com/ayoreis/import
 */

const SHEBANG = /^#!.*/;

const esbuildOptions: BuildOptions = {
  bundle: true,
  platform: "neutral",
  format: "esm",
  write: false,
  ignoreAnnotations: true,
  keepNames: true,
  treeShaking: false,
  logLevel: "error",
  plugins: denoPlugins({
    loader: "portable" as const,
  }) as Plugin[],
};

const AsyncFunction = async function () {}.constructor;

async function buildAndEvaluate(
  options: BuildOptions,
) {
  const { outputFiles } = await build(
    {
      ...esbuildOptions,
      ...options,
    },
  );

  await stop();

  if (!outputFiles) {
    throw new Error("Failed to build");
  }

  const [{ text }] = outputFiles;
  const [
    before,
    after = "}",
  ] = text.split("export {");

  const body = before.replace(SHEBANG, "") +
    "return {" +
    after.replaceAll(
      /(?<local>\w+) as (?<exported>\w+)/g,
      "$<exported>: $<local>",
    );

  const exports = await AsyncFunction(body)();

  const toStringTaggedExports = {
    [Symbol.toStringTag]: "Module",
    ...exports,
  };

  const sortedExports = Object.fromEntries(
    Object.keys(toStringTaggedExports)
      .sort()
      .map((key) => [key, toStringTaggedExports[key]]),
  );

  const emptyObject = Object.create(null);
  const prototypedExports = Object.assign(emptyObject, sortedExports);
  const sealedExports = Object.seal(prototypedExports);

  return sealedExports;
}

export async function dynamicImport(modulePath: string) {
  return await buildAndEvaluate(
    { entryPoints: [modulePath] },
  );
```
usage
```ts
     const { default: config } = await dynamicImport(path);
```


ideally this could be handled by deno.js But this work well enough from me.

**Comment by andyburke** [comment]
> I'm using webworkers to load .ts modules that are on the local machine but not known at compile-time. It's basically a user addon system. It is of course running fine with `deno run`. Hitting the code path where the import happens in the compiled version gives me `ERR_MODULE_NOT_FOUND`.

I am running up against a similar use case.

I would like to be able to dynamically import .ts files from a compiled deno app. I don't want to introduce a ton of dependencies to achieve what seems to work fine in the uncompiled setup. Is there any chance there will be progress here to allow for importing Typescript?

**Comment by vectronic** [comment]
FWIW, I switched from Deno to Bun because it supports dynamic import from a compiled bun app out of the box.

But (there is always a but...) there is no support for loading from a remote URL: https://github.com/oven-sh/bun/issues/38

So dynamic imports need reference a local filesystem URL only. They can be absolute paths or paths relative to the executable.

I wrote a little helper to pull remote modules locally:

https://github.com/flowscripter/dynamic-plugin-framework/blob/main/src/plugin_manager/util/PluginLoader.ts#L26-L47

And then I can import OK here:

https://github.com/flowscripter/dynamic-plugin-framework/blob/main/src/plugin_manager/util/PluginLoader.ts#L85

**Comment by dsherret** [comment]
Note that it is possible to import javascript files not present at compile time on the file system. You just need to provide it the correct url because the referrer for modules inside the executable is inside the executable and not the cwd.

```ts
// main.ts
import { pathToFileURL } from "node:url";
import { join } from "node:path";

const base = pathToFileURL(join(Deno.cwd(), "main.js"));
await import(new URL(Deno.args[0], base).href);
```

```
> deno compile -o scratch main.ts
> cat asdf.js
console.log("Hi");
> ./scratch asdf.js
Hi
```

Remote modules and typescript files don't work atm. Probably we'll introduce a loader API so people can load typescript and remote modules using something like https://jsr.io/@deno/loader -- That would allow not bloating the compiled executable for everyone and people could only include a typescript parser & transpiler if they need it (right now the compiled executable does not have a typescript parser or transpiler)

**Comment by davidbarratt** [comment]
I changed my mind on this issue. It's akin to a DLL and that can basically land you in dependency hell.

**Comment by 0xNF** [comment]
> Note that it is possible to import javascript files not present at compile time on the file system. You just need to provide it the correct url because the referrer for modules inside the executable is inside the executable and not the cwd.


How does this work for .wasm blobs? That's another of my use cases for this feature.

**Comment by dsherret** [comment]
It works with Wasm.

Same executable as above:

<img width="801" height="548" alt="Image" src="https://github.com/user-attachments/assets/cb38ff7d-2794-4be6-b609-54296adbe49f" />

**Comment by brad-jones** [comment]
I was noodling away with this https://github.com/brad-jones/deno-ts-importer but eventually gave up.

I wonder if there is an option where an alternative `denort` bin could be used by `deno compile` that did have the TypeScript transpiler and other bits to make dynamic Typescript imports _just work ™️_

Something like `deno compile --mode full | lite` where `lite` would be the default and use the current `denort` bin. But full would use a larger runtime bin with just enough to make imports work they way everyone here kind of thought they might.

Just an idea.

**Comment by dsherret** [comment]
The idea is that custom loaders would enable easily adding flags to `deno compile` that would then pull in the code to do a loader implementation that implements these features. That would keep denort smaller.

**Comment by bartlomieju** [comment]
For those following this issue: dynamic imports of **JavaScript files** from the filesystem already work in `deno compile` binaries today. If the file isn't embedded in the binary, the runtime falls back to reading from the real filesystem. So this works:

```ts
// main.ts (compiled into binary)
const pluginPath = Deno.args[0];
const mod = await import(`file://${Deno.realPathSync(pluginPath)}`);
mod.greet("world"); // works if plugin.js is on disk
```

The remaining gap is **TypeScript** — since the compiled runtime (`denort`) doesn't include the TypeScript transpiler (SWC), `.ts` files loaded at runtime hit V8 as raw TypeScript and fail.

With the recent `module.registerHooks()` implementation (#33733, #33763), users will soon be able to solve this in userland by registering a custom loader that transpiles TS before it reaches V8. However, this doesn't work in compiled binaries yet because the standalone module loader doesn't wire up the hook registry (`hook_registry: None`).

We're looking into enabling `module.registerHooks()` in compiled binaries, which would let users bundle their own transpiler and handle dynamic TS imports without needing SWC built into `denort`.

**Comment by 0xNF** [comment]
Thanks to all involved for solving this issue! It was a real thorn in my side for some projects. Deno compile with dynamic modules works really nicely now, it's a really nice way to implement a plugin system.
